GoogleTestAdapter/Core/Runners/CommandLineGenerator.cs

```diff
@@ -69,39 +69,59 @@ private IEnumerable<Args> GetFinalCommandLines(string baseCommandLine)
                 return commandLines;
             }
 
-            string baseCommandLineWithFilter = baseCommandLine + GoogleTestConstants.FilterOption;
-
             List<string> suitesRunningAllTests = GetSuitesRunningAllTests();
-            baseCommandLineWithFilter += GetFilterForSuitesRunningAllTests(suitesRunningAllTests);
-
-            List<TestCase> testCasesNotRunBySuite = GetTestCasesNotRunBySuite(suitesRunningAllTests);
-            List<TestCase> testCasesRunBySuite = _testCasesToRun.Where(tc => !testCasesNotRunBySuite.Contains(tc)).ToList();
-            if (testCasesNotRunBySuite.Count == 0)
+            int suiteLengthAggregate = 0;
+            int groupIdxCounter = 0;
+            int maxSuiteLength = MaxCommandLength - _lengthOfExecutableString - userParam.Length - 1;
+
+            var suiteGroups = suitesRunningAllTests
+                              .Select(suite =>
+                                { suiteLengthAggregate += suite.Length + 3;
+                                  if (suiteLengthAggregate > maxSuiteLength) {
+                                      ++groupIdxCounter;
+                                      suiteLengthAggregate = suite.Length + 3;
+                                  }
+                                  return new { groupIdx = groupIdxCounter, suiteName = suite };                              
+                                })
+                              .GroupBy(pair => pair.groupIdx, pair => pair.suiteName)
+                              .ToList();
+
+            if (!suiteGroups.Any()) suiteGroups = Enumerable.Range(-1, 1).GroupBy(dummy => dummy, dummy => string.Empty).ToList();
+
+            foreach (var suiteGroup in suiteGroups)
             {
-                commandLines.Add(new Args(_testCasesToRun, baseCommandLineWithFilter + userParam));
-                return commandLines;
-            }
-
-            List<TestCase> includedTestCases;
-            int remainingLength = MaxCommandLength
-                - baseCommandLineWithFilter.Length - _lengthOfExecutableString - userParam.Length - 1;
-            string commandLine = baseCommandLineWithFilter +
-                JoinTestsUpToMaxLength(testCasesNotRunBySuite, remainingLength, out includedTestCases);
-            includedTestCases.AddRange(testCasesRunBySuite);
-            commandLines.Add(new Args(includedTestCases, commandLine + userParam));
-
-            // only add suites to first command line
-            baseCommandLineWithFilter = baseCommandLine + GoogleTestConstants.FilterOption;
-
-            while (testCasesNotRunBySuite.Count > 0)
-            {
-                remainingLength = MaxCommandLength
+              List<string> suites = suiteGroup.Key >= 0 ? suiteGroup.ToList() : Enumerable.Empty<string>().ToList();
```

Comment:
I also do not get this, but it's probably related to the above...

Reply:
yes. the "little bit" of special handling needed. As I only have a "fake" suiteGroup (if Key < 0) I need to construct a empty list of suites (as suiteGroup.ToList() would not work properly here).

Reply:
So I moved the suite-list-creation into its own method, removed all the "special handling", introduced the "preamble" as a Func (with anonymous return type), and split the loop into the "all but the last suite-list" part and the last suite-list part (which also creates the command lines with non-suite-filters)
Hopefully this improved readability and coherence.